File: d01.py
#!/usr/bin/env python3

import util

def fuel(x, tyranny=False):
    output = max(x//3 - 2, 0)
    if output and tyranny:
        return output + fuel(output, tyranny=True)
    else:
        return output

def main():
    inputs = [int(x) for x in util.get_puzzle_input().split()]
    total_fuel = sum(fuel(x) for x in inputs)
    print(total_fuel)
    total_fuel = sum(fuel(x,tyranny=True) for x in inputs)
    print(total_fuel)


if __name__=='__main__':
    main()

File: cpp/d01.cc
#include <algorithm>
#include <iostream>
#include <string>

#include "ReadInputs.hh"

int fuel(int weight, bool tyranny=false) {
  int output = std::max(weight/3 - 2, 0);

  if(output && tyranny) {
    return output + fuel(output, true);
  } else {
    return output;
  }
}

int main(int argc, char** argv) {
  if(argc < 2) {
    std::cerr << "Missing argument: puzzle input" << std::endl;
    return 1;
  }

  std::string puzzle_input_file(argv[1]);

  auto inputs = read_integers(puzzle_input_file);

  int total_fuel = 0;
  for(auto weight : inputs) {
    total_fuel += fuel(weight);
  }
  std::cout << "Without tyranny: " << total_fuel << std::endl;

  total_fuel = 0;
  for(auto weight : inputs) {
    total_fuel += fuel(weight, true);
  }
  std::cout << "With tyranny: " << total_fuel << std::endl;
}
